# Lista de exercícios - Professor Eduardo

Observações:
- Pode-se usar softwares (e.g., R, Python, Weka etc.) para resolver as questões.
- Todos os arquivos CSV possuem cabeçalho com o nome das colunas e campos separados por vírgula “,”.
- Todas as questões têm pesos iguais e valem 2,5 pontos.
- Data máxima para entrega: 05/12/2022 no horário da aula(14h). Para cada dia de atraso na entrega serão
descontados 2,5 pontos.

Informações:
- NUSP: 11259715 Nome: Vanderson da S. dos Santos
- NUSP: 11261110 Nome: Antonio Lago Araújo Seixas 
- NUSP: 11375354 Nome: Greghory do Amaral

- Matéria: PCS3438 - Inteligência Artificial

- Professor: Eduardo Raul Hruschka  

- Data: 07/12/2022

---

Para o desenvolvimento do exercício foi escolhido a linguagem `Python` e foram utilizadas as bibilitecas `numpy`, `pandas` e `sklearn`, sendo cada uma usadas para:

- pandas: para o tratamentos dos dados disponibilizados.
- numpy: para manipulações matemáticas.
- sklearn: para o treinamento de modelos, predição, cálculo de erros e
validação cruzada.

In [28]:
# generic imports
import os                
import numpy as np
import pandas as pd

# machine learning libraries
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.metrics import accuracy_score

---

Antes de começar a resolver os exercícios, foi criado uma classe generica para fazer upload dos dados

In [29]:
class Question:
    def __init__(self,dataset_name):
        self.InputDataReader = lambda data: pd.read_csv(data,sep=r'\s*,\s*',engine='python',na_values="?")

        self.notebook_folder_path = os.getcwd()
        self.list_folder_path = f"{self.notebook_folder_path}/../../../assets/dataset/eduardo/lista"

        self.dataset_path = f"{self.list_folder_path}/{dataset_name}.csv"
        self.dataset = self.InputDataReader(self.dataset_path)

---

## Questão 1

Considerando os dados presentes no arquivo class01.csv, treine o algoritmo Naive Bayes Gaussiano utilizando a metodologia de validação cruzada holdout (utilize para treino as 350 primeiras linhas e para validação as demais). Qual o valor da acurácia a base de treino? Qual o valor da acurácia na base de validação?

Seguindo as instruções informadas no exercício, foi realizado o treinamento do algoritmo Naives Bayes Gaussiano utilizando a validação cruzada holdout. 

In [30]:
from sklearn.naive_bayes import GaussianNB

In [31]:
class Question1(Question):
    def __init__(self):
        super().__init__("class01")

question1 = Question1()
# question1.dataset.head(5)

In [32]:
def prepare_data_question1(self):
    self.train_data, self.validation_data = train_test_split(self.dataset, test_size=0.65)
    self.validation_data_without_target = self.validation_data.drop(["target"], axis=1)
    self.train_data_without_target = self.train_data.drop(["target"], axis=1)

    print(f"Tamanho dados de treino : {len(self.train_data)}\nTamanho dados de validação : {len(self.validation_data)}") 

Question1.prepare_data = prepare_data_question1
question1.prepare_data()

Tamanho dados de treino : 350
Tamanho dados de validação : 650


Podemos ver aqui que houveram 350 dados de treino e 650 de validação

In [33]:
def run_question1(self):
    self.model = GaussianNB()
    
    self.train_data_labs = self.train_data.columns.values
    self.train_data_labs = self.train_data_labs[:-1]

    self.train_data_x = self.train_data[self.train_data_labs]
    self.train_data_y = self.train_data[["target"]]


    # cross_val = cross_val_score(self.model, self.train_data_x,self.train_data_y.values.ravel(), cv=10,error_score='raise')
    self.model.fit(self.train_data_x,self.train_data_y.values.ravel())

    self.validation_data_predicted = self.model.predict(self.validation_data_without_target)
    self.train_data_predicted = self.model.predict(self.train_data_without_target)

Question1.run = run_question1
question1.run()

In [34]:
def validation_question1(self):
    self.validation_data_y = self.validation_data[["target"]]

    self.validation_accuracy_score = accuracy_score(self.validation_data_y,self.validation_data_predicted)
    self.train_accuracy_score = accuracy_score(self.train_data_y,self.train_data_predicted)

Question1.validation = validation_question1
question1.validation()

In [35]:
def show_results_question1(self):
    msg = f"A acurácia da base de treino foi de {self.train_accuracy_score*100} \nA acurácia da base de validação foi de {self.validation_accuracy_score*100}"
    print(msg)

Question1.show_results = show_results_question1
question1.show_results()

A acurácia da base de treino foi de 77.14285714285715 
A acurácia da base de validação foi de 61.69230769230769


### Questão 1 - Resultado

Após o calculos, chegamos ao resultado de:

- A acurácia da base de treino foi de 77.14%
- A acurácia da base de validação foi de 61.69%

---

## Questão 2

Considerando os dados presentes no arquivo class02.csv, treine o algoritmo 10-Nearest Neighbors (KNN com 𝑘 = 10 e distância Euclidiana), utilizando a metodologia de validação cruzada k-fold com 5 folds. Considere que a primeira pasta de validação seja formada pelas primeiras 20% linhas do arquivo, que a segunda pasta de validação seja formada pelas 20% linhas seguintes, e assim por diante, até atingir a última pasta, formada pelas 20% linhas finais da base. Qual a acurácia média para a base de validação?

Como informado no enunciado, foi realizado o treinamento do algoritmo 10-Nearest Neighbors, com k = 10 de distância Euclidiana, utilizando a validação cruzada descrita. Sua implementação pode ser vista no código abaixo.

In [36]:
from sklearn.neighbors import KNeighborsClassifier 

In [37]:
class Question2(Question):
    def __init__(self):
        super().__init__("class02")

question2 = Question2()
# question2.dataset.head(5)

In [38]:
def prepare_data_question2(self):
    self.train_data, self.validation_data = train_test_split(self.dataset, test_size=0.30) 
    self.validation_data_without_target = self.validation_data.drop(["target"], axis=1)
    self.train_data_without_target = self.train_data.drop(["target"], axis=1)

    print(f"Tamanho dados de treino : {len(self.train_data)}\nTamanho dados de validação : {len(self.validation_data)}") 

Question2.prepare_data = prepare_data_question2
question2.prepare_data()

Tamanho dados de treino : 1050
Tamanho dados de validação : 450


A validação cruzada k-fold (com 5 folds) funciona um pouco diferente da primeira validação cruzada

<img src="../../../assets/docs/cross_validation_k_fold.png" width="500" />

Sendo que normalmente nesse tipo de validação cruzada os dados de validação pegam 30% de todo o dataset e os de treino 70%

In [39]:
def run_question2(self):
    self.model = KNeighborsClassifier(n_neighbors=5)
    
    self.train_data_labs = self.train_data.columns.values
    self.train_data_labs = self.train_data_labs[:-1]

    self.train_data_x = self.train_data[self.train_data_labs]
    self.train_data_y = self.train_data[["target"]]

    self.k_fold = 5
    cross_val = cross_val_score(self.model, self.train_data_x,self.train_data_y.values.ravel(), cv=self.k_fold,error_score='raise')
    self.model.fit(self.train_data_x,self.train_data_y.values.ravel())

    self.validation_data_predicted = self.model.predict(self.validation_data_without_target)
    self.train_data_predicted = self.model.predict(self.train_data_without_target)

    print(f"Cross validation result(R² with k-fold = {self.k_fold}) = {cross_val}")

Question2.run = run_question2
question2.run()

Cross validation result(R² with k-fold = 5) = [0.83809524 0.77142857 0.84285714 0.81904762 0.78571429]


In [40]:
def validation_question2(self):
    self.validation_data_y = self.validation_data[["target"]]

    self.validation_accuracy_score = accuracy_score(self.validation_data_y,self.validation_data_predicted)
    self.train_accuracy_score = accuracy_score(self.train_data_y,self.train_data_predicted)

Question2.validation = validation_question2
question2.validation()

In [41]:
def show_results_question2(self):
    msg = f"A acurácia da base de treino foi de {self.train_accuracy_score*100} \nA acurácia da base de validação foi de {self.validation_accuracy_score*100}"
    print(msg)

Question2.show_results = show_results_question2
question2.show_results()

A acurácia da base de treino foi de 88.38095238095238 
A acurácia da base de validação foi de 81.77777777777779


### Questão 2 - Resultados

Após o calculos, chegamos ao resultado de:

- A acurácia da base de treino foi de 88.38% 
- A acurácia da base de validação foi de 81.78%

Como pode ser observado, houve um desempenho para o modelo adotado

---

## Questão 3

Considerando os dados presentes no arquivo reg01.csv, obtenha um modelo de regressão linear com regularização L1 (LASSO com 𝛼 = 1) utilizando a metodologia Leave-One-out. Qual o valor médio do Root Mean Squared Error (RMSE) para a base de treino e para a base de validação? 

Segue abaixo a implementação do código para a resolução do exercício.

In [42]:
from sklearn.linear_model import Lasso
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import LeaveOneOut

In [43]:
class Question3(Question):
    def __init__(self):
        super().__init__("reg01")

question3 = Question3()
# question3.dataset.head(5)

In [44]:
def prepare_data_question3(self):
    self.train_data, self.validation_data = train_test_split(self.dataset, test_size=0.30) 
    self.validation_data_without_target = self.validation_data.drop(["target"], axis=1)
    self.train_data_without_target = self.train_data.drop(["target"], axis=1)

    print(f"Tamanho dados de treino : {len(self.train_data)}\nTamanho dados de validação : {len(self.validation_data)}") 

Question3.prepare_data = prepare_data_question3
question3.prepare_data()

Tamanho dados de treino : 700
Tamanho dados de validação : 300


In [45]:
def run_question3(self):
    self.model = Lasso(alpha=1)
    
    self.train_data_labs = self.train_data.columns.values
    self.train_data_labs = self.train_data_labs[:-1]

    self.train_data_x = self.train_data[self.train_data_labs]
    self.train_data_y = self.train_data[["target"]]

    self.cv = LeaveOneOut()
    self.cross_val = cross_val_score(self.model, self.train_data_x,self.train_data_y.values.ravel(), cv=self.cv, scoring='neg_mean_absolute_error', n_jobs=-1)
    
    self.model.fit(self.train_data_x,self.train_data_y.values.ravel())

    self.validation_data_predicted = self.model.predict(self.validation_data_without_target)
    self.train_data_predicted = self.model.predict(self.train_data_without_target)

Question3.run = run_question3
question3.run()

Para calcular o Root Mean Squared Error (RMSE), devemos seguir a seguinte formula:

$ RMSE = \sqrt{\frac{\sum^{N}_{i}{(yi_{predict} - yi_{actual})^{2}}}{N}} $

Para facilitar, temos uma função do sklearn que faz isso por nós

In [46]:
def validation_question3(self):
    self.validation_data_y = self.validation_data[["target"]]
    
    self.validation_data_rmse = np.sqrt(mean_squared_error(self.validation_data_y,self.validation_data_predicted))
    self.train_data_rmse = np.sqrt(mean_squared_error(self.train_data_y,self.train_data_predicted))

Question3.validation = validation_question3
question3.validation()

In [47]:
def show_results_question3(self):
    msg = f"A RMSE da base de treino foi de {self.train_data_rmse} \nA RMSE da base de validação foi de {self.validation_data_rmse}"
    print(msg)

Question3.show_results = show_results_question3
question3.show_results()

A RMSE da base de treino foi de 19.582348396640022 
A RMSE da base de validação foi de 18.541233298227333


### Questão 3 - Resposta

Após o calculos, chegamos ao resultado de:

- A RMSE da base de treino foi de 19.582348396640022 
- A RMSE da base de validação foi de 18.541233298227333

---

## Questão 4

Considerando os dados do arquivo reg02.csv, treine árvores de regressão, sem realizar podas, utilizando a metodologia de validação cruzada k-fold com 𝑘 = 5. Qual o valor do Mean Absolute Error (MAE) para a base de treino? Qual o valor médio do MAE para a base de validação? 

In [48]:
from sklearn.tree import DecisionTreeRegressor

In [49]:
class Question4(Question):
    def __init__(self):
        super().__init__("reg02")

question4 = Question4()
# question4.dataset.head(5)

In [50]:
def prepare_data_question4(self):
    self.train_data, self.validation_data = train_test_split(self.dataset, test_size=0.30) 
    self.validation_data_without_target = self.validation_data.drop(["target"], axis=1)
    self.train_data_without_target = self.train_data.drop(["target"], axis=1)

    print(f"Tamanho dados de treino : {len(self.train_data)}\nTamanho dados de validação : {len(self.validation_data)}") 

Question4.prepare_data = prepare_data_question4
question4.prepare_data()

Tamanho dados de treino : 700
Tamanho dados de validação : 300


Para calcular o MAE, usamos a seguinte formula:

$ MAE = \frac{\sum^{N}_{i}{|(yi_{predict} - yi_{actual})|}}{N} $

In [51]:
def run_question4(self):
    self.model = DecisionTreeRegressor()
    
    self.train_data_labs = self.train_data.columns.values
    self.train_data_labs = self.train_data_labs[:-1]

    self.train_data_x = self.train_data[self.train_data_labs]
    self.train_data_y = self.train_data[["target"]]

    self.k_fold = 5
    cross_val = cross_val_score(self.model, self.train_data_x,self.train_data_y.values.ravel(), cv=self.k_fold,error_score='raise')
    self.model.fit(self.train_data_x,self.train_data_y.values.ravel())

    self.validation_data_predicted = self.model.predict(self.validation_data_without_target)
    self.train_data_predicted = self.model.predict(self.train_data_without_target)

    print(f"Cross validation result(R² with k-fold = {self.k_fold}) = {cross_val}")

Question4.run = run_question4
question4.run()

Cross validation result(R² with k-fold = 5) = [0.97822732 0.98533619 0.98190539 0.97937487 0.98187906]


In [52]:
def validation_question4(self):
    self.validation_data_y = self.validation_data[["target"]]
    
    self.calculation_mae = lambda data_true, data_predicted: np.mean(np.abs(data_true - data_predicted))

    self.validation_data_mae = self.calculation_mae(self.validation_data_y.values.ravel(),self.validation_data_predicted)
    self.train_data_mae = self.calculation_mae(self.train_data_y.values.ravel(),self.train_data_predicted) 

Question4.validation = validation_question4
question4.validation()

In [53]:
def show_results_question4(self):
    msg = f"A MAE da base de treino foi de {self.train_data_mae} \nA MAE da base de validação foi de {self.validation_data_mae}"
    print(msg)

Question4.show_results = show_results_question4
question4.show_results()

A MAE da base de treino foi de 0.0 
A MAE da base de validação foi de 45.42253496774727


### Questão 4 - Resposta

Após o calculos, chegamos ao resultado de:

- A MAE da base de treino foi de 0.0 
- A MAE da base de validação foi de 45.994092472635536

---